### PR TITLE
More API price formatting and currency handling

### PR DESCRIPTION
**Pull request by mikejolley**
This took a lot longer than I'd hoped, but this should now be in a better place.

This PR introduces a consistent response format for endpoints returning prices/totals via a new method `get_store_currency_properties` and `get_store_currency_response`. When prices are returned, currency and minor unit information is also returned. This mostly affects product and cart endpoints.

As well as this, other endpoints handling prices have been updated. Price filtering (and active price filters), and product price templates being the main ones. These now also deal with subunit/cent values, and formatting is taken care of using the `react-number-format` library where possible.

### How to test the changes in this Pull Request:

1. Check API responses for /cart/ `totals` and product `prices` show correct pricing data and return the currency.
2. Test that you can filter by price using the all products block, and that prices are shown in the correct format.
3. Check using other currencies in WooCommerce, and/or change the number of decimals in settings.

Next steps:
- After merge, update coupons PR to follow these patterns.
- Add Todos for things which could use improvement. I noticed the CURRENCY setting uses `precision` and other terminology which could better match the API. Additionally, using prefix/suffix for currency seems to be easier to handle that sprintf/`priceFormat` as we have in places.

### CODE REVIEW

File: assets/js/base/components/price-slider/index.js [hunk 1]
```diff
@@ -187,15 +189,24 @@ const PriceSlider = ( {
 	 */
 	const priceInputOnBlur = useCallback(
 		( event ) => {
+			// Only refresh when finished editing the min and max fields.
```
Comment (mikejolley) [inline assets/js/base/components/price-slider/index.js]: I added this block to prevent queries running when tabbing between the two pricing fields which I noticed when testing!

File: assets/js/base/components/price-slider/index.js [hunk 2]
```diff
@@ -300,20 +280,82 @@ const PriceSlider = ( {
...
-						maxPrice={ formattedMaxPrice }
-					/>
-				) : (
-					<PriceLabel
-						minPrice={ formattedMinPrice }
-						maxPrice={ formattedMaxPrice }
-					/>
+				{ showInputFields && (
+					<Fragment>
+						<NumberFormat
```
Comment (mikejolley) [inline assets/js/base/components/price-slider/index.js]: Using NumberFormat here for inputs and labels seemed more efficient than storing custom price strings and passing them around.

File: assets/js/base/utils/price.js [hunk 3]
```diff
@@ -5,25 +5,21 @@ import { sprintf } from '@wordpress/i18n';
...
 		return '';
 	}
-	const formattedNumber = parseInt( value, 10 );
+
+	// eslint-disable-next-line @wordpress/valid-sprintf
 	const formattedValue = sprintf(
-		priceFormat,
-		currencySymbol,
-		formattedNumber
+		CURRENCY.priceFormat,
```
Comment (mikejolley) [inline assets/js/base/utils/price.js]: We may want to add or rename some settings from this object. For now I've just updated to use what we have already defined.

File: assets/js/base/utils/price.js [hunk 4]
```diff
@@ -5,25 +5,21 @@ import { sprintf } from '@wordpress/i18n';
...
+		CURRENCY.symbol,
+		parseInt( price, 10 ) / 10 ** CURRENCY.precision
```
Comment (mikejolley) [inline assets/js/base/utils/price.js]: Just a note because I wasn't aware until today, pow( 0 ) = 1, so we shouldn't see division by 0 errors.

File: assets/js/blocks/active-filters/block.js [hunk 5]
```diff
@@ -34,7 +34,7 @@ const ActiveFiltersBlock = ( {
 			return null;
 		}
 		return renderRemovableListItem(
-			__( 'Price:', 'woo-gutenberg-products-block' ),
+			__( 'Price', 'woo-gutenberg-products-block' ),
```
Comment (mikejolley) [inline assets/js/blocks/active-filters/block.js]: There was a double :: here missed in testing.

File: src/RestApi/StoreApi/Schemas/CartItemSchema.php [hunk 6]
```diff
@@ -187,19 +192,23 @@ public function get_item_response( $cart_item ) {
...
 
 		return [
-			'key'               => $cart_item['key'],
-			'id'                => $product->get_id(),
-			'quantity'          => wc_stock_amount( $cart_item['quantity'] ),
-			'name'              => $product->get_title(),
-			'sku'               => $product->get_sku(),
-			'permalink'         => $product->get_permalink(),
-			'images'            => ( new ProductImages() )->images_to_array( $product ),
-			'product_price'     => $this->prepare_money_response( $product->get_price(), wc_get_price_decimals() ),
```
Comment (mikejolley) [inline src/RestApi/StoreApi/Schemas/CartItemSchema.php]: Note; I removed this intentionally as it isn't used. Cart should only need the line totals below.

**Comment by mikejolley** [comment]
After merging from master Prettier kicked in and changed some stuff (https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1394/commits/63ab30c6c62262afb370d7b5120fbd70282e6d08) -.- Not sure why. @Aljullu You had similar problems?

**Comment by senadir** [comment]
> After merging from master Prettier kicked in and changed some stuff ([63ab30c](https://github.com/woocommerce/woocommerce-gutenberg-products-block/commit/63ab30c6c62262afb370d7b5120fbd70282e6d08)) -.- Not sure why. @Aljullu You had similar problems?

This happens all the time when I merge from master or commit changes

**Comment by Aljullu** [comment]
> After merging from master Prettier kicked in and changed some stuff (63ab30c) -.- Not sure why. @Aljullu You had similar problems?

I didn't but I'm a bit confused with it. When I run `npm run reformat-files` in this branch I get several more changes:

https://gist.github.com/Aljullu/2ac90ed4b068a6bc721619ece5e5d0aa

Can you reproduce that too? I have `wp-prettier` 1.91.1 and it uses the `.prettierrc` config file from this repo.

> This happens all the time when I merge from master or commit changes

Right, we have Prettier configured with lint-staged and it runs after the commit. So formatted files must be added after commit, which might be easy to miss.

What I think works best is installing Prettier in your editor (https://prettier.io/docs/en/editors.html) and setting it to format on save (VS Code: https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode#format-on-save).

File: assets/js/atomic/components/product/price/index.js [hunk 7]
```diff
@@ -33,12 +34,18 @@ const ProductPrice = ( { className, product } ) => {
 					className={ `${ layoutStyleClassPrefix }__product-price__value` }
 				>
 					<NumberFormat
-						value={ prices.price_range.min_amount }
+						value={
+							prices.price_range.min_amount /
+							10 ** prices.currency_minor_unit
```
Comment (Aljullu) [inline assets/js/atomic/components/product/price/index.js]: `x / 10 ** prices.currency_minor_unit` seems to be used in several places in this and other files of this PR. I was wondering if it might make sense extracting it to a util function? I don't have a strong opinion about this so mostly asking what you think, but I find operations with exponents are a bit hard to rationale if you are not used to the code, so I think encapsulating it into its own function might make the code easier to read.

**Comment by mikejolley** [comment]
@Aljullu I have it on save too, but I suspect this occurs because when merging from master there are changes needed in files I do not touch?

File: assets/js/base/hocs/test/with-products.js.rej [hunk 8]
```diff
@@ -0,0 +1,7 @@
+diff a/assets/js/base/hocs/test/with-products.js b/assets/js/base/hocs/test/with-products.js	(rejected hunks)
+@@ -21 +21,4 @@ jest.mock( '../../utils/errors', () => ( {
+-const mockProducts = [ { id: 10, name: 'foo' }, { id: 20, name: 'bar' } ];
++const mockProducts = [
++	{ id: 10, name: 'foo' },
++	{ id: 20, name: 'bar' },
++];
```
Comment (Aljullu) [inline assets/js/base/hocs/test/with-products.js.rej]: We must remember to remove this file.

**Comment by nerrad** [comment]
> but I suspect this occurs because when merging from master there are changes needed in files I do not touch?

7 days ago we updated `wp-prettier` which includes a number of changes to make it a bit more compatible with the WordPress js standards.  My guess is that these changes we are seeing are exactly that, files that have formatting impacts from the update.

**Comment by nerrad** [comment]
Also, if you are _merging_ master in, vs rebasing on top of master, that will impact what prettier touches.  It's another reason to avoid merging instead of rebasing.  Personally, I'd still like to see us always rebase for pulls but I know you guys seem to prefer merging 🤷‍♂️

File: assets/js/blocks/active-filters/block.js (hunk 5 above)
Comment (Aljullu) [reply assets/js/blocks/active-filters/block.js]: Should we create a PR with this change targeting 2.5.5? That would avoid translators to have to translate this twice.

File: src/RestApi/StoreApi/README.md [hunk 9]
```diff
@@ -121,8 +121,11 @@ Example response:
 
 ```json
 {
-	"min_price": "0.00",
-	"max_price": "90.00",
+	"price_range": [
+    "currency_minor_unit": 2,
```
Comment (Aljullu) [inline src/RestApi/StoreApi/README.md]: There are some alignment issues here.

File: assets/js/base/components/price-slider/index.js (hunk 1 above)
Comment (nerrad) [reply assets/js/base/components/price-slider/index.js]: good catch.

File: assets/js/base/components/price-slider/index.js [hunk 10]
```diff
@@ -187,15 +189,24 @@ const PriceSlider = ( {
...
-			const targetValue = event.target.value.replace( /[^0-9.-]+/g, '' );
-			const currentValues = isMin
-				? [ targetValue, maxPrice ]
-				: [ minPrice, targetValue ];
 			const values = constrainRangeSliderValues(
-				currentValues,
+				[
+					minPriceInput * 10 ** decimalScale,
+					maxPriceInput * 10 ** decimalScale,
+				],
```
Comment (nerrad) [inline assets/js/base/components/price-slider/index.js]: You should be able to just pass the current values here right?  I think as much as possible we should do all calculations/comparisons on the subunit values and only do conversion to currency representation with formatting on display.

File: assets/js/base/components/price-slider/index.js (hunk 2 above)
Comment (nerrad) [reply assets/js/base/components/price-slider/index.js]: Agreed.  If we implement the `FormattedPrice` price component idea I had, it could have custom events internally that automatically convert back and forth from the subunit values so your callbacks as implemented here deal with subunits (which keeps things cleanly separated from display and internal representation).  This will also have importance for potential interfaces where components using `FormattedPrice` inputs could be sending back data to the server over the wire (where values need to be subunits).

File: assets/js/base/utils/price.js [hunk 11]
```diff
@@ -5,25 +5,21 @@ import { sprintf } from '@wordpress/i18n';
...
- * @param {number} value Number to format.
- * @param {string} priceFormat  Price format string.
- * @param {string} currencySymbol Currency symbol.
+ * @param {number} price Price, in cents, to format.
  */
-export const formatPrice = (
-	value,
-	priceFormat = CURRENCY.priceFormat,
-	currencySymbol = CURRENCY.symbol
-) => {
```
Comment (nerrad) [inline assets/js/base/utils/price.js]: Why this change?  Originally this format helper defaulted to set currency values from the server but still allowed for overriding for specific formats.  Now this ONLY uses currency values from the server.  This change also breaks tests for the utility function.

File: assets/js/blocks/price-filter/block.js [hunk 12]
```diff
@@ -36,9 +37,24 @@ const PriceFilterBlock = ( { attributes, isEditor = false } ) => {
...
+					currency_code: 'USD',
+					currency_symbol: '$',
+					currency_minor_unit: 2,
+					currency_decimal_separator: '.',
+					currency_thousand_separator: ',',
+					currency_prefix: '$',
+					currency_suffix: '',
+					min_price: undefined,
+					max_price: undefined,
+			  }
```
Comment (nerrad) [inline assets/js/blocks/price-filter/block.js]: This fallback should be defined outside the component so that the object is only created once.

File: assets/js/blocks/price-filter/block.js [hunk 13]
```diff
@@ -116,9 +132,16 @@ const PriceFilterBlock = ( { attributes, isEditor = false } ) => {
 					maxConstraint={ maxConstraint }
 					minPrice={ min }
 					maxPrice={ max }
-					step={ 10 }
-					currencySymbol={ CURRENCY.symbol }
-					priceFormat={ CURRENCY.priceFormat }
+					step={ 10 ** priceRangeData.currency_minor_unit }
```
Comment (nerrad) [inline assets/js/blocks/price-filter/block.js]: I think all currency and calcs should be handled by default in the PriceSlider if the props are not provided.   That will eliminate the need for any currency information inside this `PriceFilterBlock` component and values can remain stored internally as subunits.

File: src/RestApi/StoreApi/Schemas/AbstractSchema.php [hunk 14]
```diff
@@ -64,6 +64,95 @@ function( $property ) {
...
+			case 'left_space':
+				$prefix = $symbol . ' ';
+				break;
+			case 'left':
+				$prefix = $symbol;
+				break;
+			case 'right_space':
+				$suffix = ' ' . $symbol;
+				break;
+			case 'right':
```
Comment (nerrad) [inline src/RestApi/StoreApi/Schemas/AbstractSchema.php]: These cases would be best defined as constants.  I realize this is likely a Woo Core thing though so no need to block this pull.

File: src/RestApi/StoreApi/Schemas/CartItemSchema.php (hunk 6 above)
Comment (nerrad) [reply src/RestApi/StoreApi/Schemas/CartItemSchema.php]: Documentation should be updated.

So is `line_subtotal` equivalent to what `product_price` was?

File: src/RestApi/StoreApi/Schemas/CartSchema.php [hunk 15]
```diff
@@ -54,106 +39,117 @@ protected function get_properties() {
...
+							'context'     => [ 'view', 'edit' ],
+							'readonly'    => true,
+						],
+						'total_tax'          => [
+							'description' => __( 'Total tax applied to items and shipping.', 'woo-gutenberg-products-block' ),
+							'type'        => 'string',
+							'context'     => [ 'view', 'edit' ],
+							'readonly'    => true,
+						],
+						'total_price'        => [
```
Comment (nerrad) [inline src/RestApi/StoreApi/Schemas/CartSchema.php]: Is this a common name used for this value throughout woo and it's api?  Would this be equivalent to `grand_total` and would that be a clearer term to use?

File: assets/js/atomic/components/product/price/index.js (hunk 7 above)
Comment (nerrad) [reply assets/js/atomic/components/product/price/index.js]: Another advantage of moving things into the `FormattedPrice` component is that it limits the surface area of `NumberFormat` in case we decide down the road that we want to switch to another library because of size or compatibility issues.

Comment (nerrad) [reply assets/js/atomic/components/product/price/index.js]: Instead of `FormattedPrice` it might be better to call it `FormattedMoney` as that's a bit more broader description of it's purpose. That would leave the door open for having various other implementations of `NumberFormat` for other things like `FormattedPhoneNumber` or `FormattedCreditCard` etc (which `NumberFormat` seems to be great for).

File: assets/js/base/utils/price.js (hunk 11 above)
Comment (mikejolley) [reply assets/js/base/utils/price.js]: Pretty sure the tests were only added after I changed this, before I rebased to master.

I changed it because the only usage of this util needed server defined values, and since API endpoints handling money all use server values I didn't see much benefit to allowing overrides.

File: src/RestApi/StoreApi/Schemas/CartSchema.php (hunk 15 above)
Comment (mikejolley) [reply src/RestApi/StoreApi/Schemas/CartSchema.php]: Urgh looks like my docs updates were removed after rebase. fml. Will resubmit.

`grand_total` is not used anywhere in core either. `total_price` follows the pattern of the other totals defined here.

Comment (mikejolley) [reply src/RestApi/StoreApi/Schemas/CartSchema.php]: Actually no, I was in wrong branch. The inline docs have been updated - see the readme.md file. @nerrad 

`line_subtotal` is not equivalent because it's for the line, not a single item. But it is all we need. Product price can come from products API.

File: assets/js/base/utils/price.js (hunk 11 above)
Comment (nerrad) [reply assets/js/base/utils/price.js]: I'll let you make the call.  It's not a publicly exposed api, so either way it should be fine.  Ya I think the tests were added in a pull Albert did fairly recently. Not much involved in updating them :).

File: src/RestApi/StoreApi/Schemas/CartSchema.php (hunk 15 above)
Comment (nerrad) [reply src/RestApi/StoreApi/Schemas/CartSchema.php]: > Product price can come from products API.

does that mean an additional request though (or would that be necessary anyways to get the product details for the items in the cart)?

File: src/RestApi/StoreApi/Schemas/CartItemSchema.php (hunk 6 above)
Comment (mikejolley) [reply src/RestApi/StoreApi/Schemas/CartItemSchema.php]: Replied https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1394#discussion_r358787705

Readme is updated already. No usage of this yet.

File: src/RestApi/StoreApi/Schemas/AbstractSchema.php (hunk 14 above)
Comment (mikejolley) [reply src/RestApi/StoreApi/Schemas/AbstractSchema.php]: Adding todo

File: src/RestApi/StoreApi/Schemas/CartSchema.php (hunk 15 above)
Comment (mikejolley) [reply src/RestApi/StoreApi/Schemas/CartSchema.php]: @nerrad yes it would, however, it's not needed for the cart and checkout :) Cart and checkout only deals with totals, even for displaying items.

**Comment by mikejolley** [comment]
@Aljullu This resolves the step issue you reported https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1394/commits/801140a7bce62680a05735f3b8fb01d0344f47cb Need to adjust the division to handle subunits, which is taken from the API response.

**Comment by mikejolley** [comment]
@Aljullu Not sure about your grouped product. I see the correct string: `£18 — £45`.

**Comment by mikejolley** [comment]
@nerrad I've incorporated some of that feedback to ensure we handle currency and subunit consistently. I have no implemented context as part of this PR, but added functions for dealing with currency info in API requests.

@Aljullu I've having test fails in the use-price-constraints tests and I cannot fathom why. Would you be able to take a peek?

File: assets/js/base/components/formatted-price/index.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+import { currencyToNumberFormat } from './utils';
+
+/**
+ * Formatted price component.
+ *
+ * Takes a price and returns a formatted price using the NumberFormat component.
+ *
+ * @param {Object} props Component props.
+ */
+const FormattedPrice = ( { value, currency, onValueChange, ...props } ) => {
```
[the remark this answers is not in the document]
Comment (joshuatf) [reply assets/js/base/components/formatted-price/index.js]: Agree with @nerrad that the naming here could be more open-ended, but `FormattedMoney` may also be confusing.  I think `FormattedMonetaryValue` would be accurate, but that's a mouthful.

On another note, would it be possible to pull the store's existing value for `currency` in case a value isn't passed?  Thinking this may be less error-prone for future consumers of this component.

File: assets/js/base/components/formatted-price/index.js [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+		...props,
+		...currencyToNumberFormat( currency ),
+		value: undefined,
+		currency: undefined,
+		onValueChange: undefined,
+	};
+
+	const onValueChangeWrapper = onValueChange
+		? ( values ) => {
+				const centValue = values.value * 10 ** currency.minorUnit;
```
[the remark this answers is not in the document]
Comment (joshuatf) [reply assets/js/base/components/formatted-price/index.js]: I think that makes cents :trollface: 

Possibly `subunitValue` or `minorUnitValue`.

**Comment by Aljullu** [comment]
> @Aljullu I've having test fails in the use-price-constraints tests and I cannot fathom why. Would you be able to take a peek?

@mikejolley the issue was that prices set in the test file were too small now that they are rounded. `minorUnit` was `2` and prices only had two digits, so they were always rounded to 0. Should be fixed in 0ba237d7b7134f60b68267cf1027d9e614c7b102.

File: assets/js/base/components/formatted-price/index.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/formatted-price/index.js]: > On another note, would it be possible to pull the store's existing value for currency in case a value isn't passed?

This happens if no currency is provided.

[the remark this answers is not in the document]
Comment (mikejolley) [reply assets/js/base/components/formatted-price/index.js]: `MonetaryAmount` is a term I've seen used in places such as schema.org. Running with that.

**Comment by mikejolley** [comment]
Made those changes and tests are now passing. Going ahead and merging this so I can start using it for coupons :) Thanks all.
